engine/src/flutter/impeller/entity/geometry/rect_geometry.cc

```diff
@@ -27,20 +28,191 @@ GeometryResult RectGeometry::GetPositionBuffer(const ContentContext& renderer,
   };
 }
 
-std::optional<Rect> RectGeometry::GetCoverage(const Matrix& transform) const {
+std::optional<Rect> FillRectGeometry::GetCoverage(
+    const Matrix& transform) const {
   return rect_.TransformBounds(transform);
 }
 
-bool RectGeometry::CoversArea(const Matrix& transform, const Rect& rect) const {
+bool FillRectGeometry::CoversArea(const Matrix& transform,
+                                  const Rect& rect) const {
   if (!transform.IsTranslationScaleOnly()) {
     return false;
   }
   Rect coverage = rect_.TransformBounds(transform);
   return coverage.Contains(rect);
 }
 
-bool RectGeometry::IsAxisAlignedRect() const {
+bool FillRectGeometry::IsAxisAlignedRect() const {
   return true;
 }
 
+StrokeRectGeometry::StrokeRectGeometry(Rect rect,
+                                       Scalar stroke_width,
+                                       Join stroke_join,
+                                       Scalar miter_limit)
+    : rect_(rect),
+      stroke_width_(stroke_width),
+      stroke_join_(AdjustStrokeJoin(stroke_join, miter_limit)) {}
+
+StrokeRectGeometry::~StrokeRectGeometry() = default;
+
+GeometryResult StrokeRectGeometry::GetPositionBuffer(
+    const ContentContext& renderer,
+    const Entity& entity,
+    RenderPass& pass) const {
+  if (stroke_width_ < 0.0) {
+    return {};
+  }
+  Scalar max_basis = entity.GetTransform().GetMaxBasisLengthXY();
+  if (max_basis == 0) {
+    return {};
+  }
+
+  Scalar min_size = kMinStrokeSize / max_basis;
+  Scalar half_stroke_width = std::max(stroke_width_, min_size) * 0.5f;
+
+  auto& host_buffer = renderer.GetTransientsBuffer();
+  Scalar left = rect_.GetLeft();
+  Scalar top = rect_.GetTop();
+  Scalar right = rect_.GetRight();
+  Scalar bottom = rect_.GetBottom();
+
+  switch (stroke_join_) {
+    case Join::kRound: {
+      Scalar radius = half_stroke_width;
+      Tessellator::Trigs trigs =
+          renderer.GetTessellator().GetTrigsForDeviceRadius(radius * max_basis);
+
+      FML_DCHECK(trigs.size() >= 2u);
+
+      // We use all but the first entry in trigs for each corner.
+      auto needed = trigs.size() - 1;
+      // Every other point has a center vertex added.
+      needed = needed + (needed >> 1);
+      // The loop also adds 3 points of its own.
+      needed += 3;
+      // We do that for each of the 4 corners.
+      needed = needed * 4;
+      // We then add 2 more points at the end to close the last edge.
+      needed += 2;
+      std::vector<Point> points;
```

Comment:
If you know the number of needed points, prefer allocating a device buffer and write directly into that instead.

Reply:
Done